Mathlib/Algebra/Category/Ring/Constructions.lean

```diff
@@ -125,6 +126,28 @@ lemma isPushout_of_isPushout (R S A B : Type u) [CommRing R] [CommRing S]
     (Algebra.IsPushout.equiv R S A B).toCommRingCatIso (by simp) (by simp)
     (by ext; simp [Algebra.IsPushout.equiv_tmul]) (by ext; simp [Algebra.IsPushout.equiv_tmul])
 
+lemma closure_range_union_range_eq_top_of_isPushout
+    {R A B X : CommRingCat.{u}} {f : R ⟶ A} {g : R ⟶ B} {a : A ⟶ X} {b : B ⟶ X}
+    (H : IsPushout f g a b) :
+    Subring.closure (Set.range a ∪ Set.range b) = ⊤ := by
+  algebraize [f.hom, g.hom]
+  let e := ((isPushout_tensorProduct R A B).isoIsPushout _ _ H).commRingCatIsoToRingEquiv
+  rw [← Subring.comap_map_eq_self_of_injective e.symm.injective (.closure _), RingHom.map_closure,
+    ← top_le_iff, ← Subring.map_le_iff_le_comap, Set.image_union]
+  simp only [AlgHom.toRingHom_eq_coe, ← Set.range_comp, ← RingHom.coe_comp]
+  rw [← hom_comp, ← hom_comp, IsPushout.inl_isoIsPushout_inv, IsPushout.inr_isoIsPushout_inv,
+    hom_ofHom, hom_ofHom]
+  rintro x -
+  induction x with
+  | zero => exact zero_mem _
+  | tmul x y =>
+    convert_to (Algebra.TensorProduct.includeLeftRingHom (R := R) x) *
+      (Algebra.TensorProduct.includeRight y) ∈ _
+    · simp
+    · exact mul_mem (Subring.subset_closure (.inl ⟨x, rfl⟩))
+        (Subring.subset_closure (.inr ⟨_, rfl⟩))
+  | add x y _ _ => exact add_mem ‹_› ‹_›
```

Comment:
It strikes me as odd that the hypotheses of the theorem are completely category-theoretic and the conclusion is completely category-free (and indeed can't really be expressed in a category-theoretic way). Is this just an inevitable consequence of having two ways to say things?

Reply:
I extracted the non-categorical part into its own lemma.